Edit pass on REST tutorial (#23942)
dotnet/docs: docs/csharp/tutorials/console-webapiclient.md

---
title: "Tutorial: Make HTTP requests in a .NET console application using C#"
description: Learn how to make HTTP requests to a REST web service and deserialize JSON responses. This tutorial creates a .NET console and uses C#.
ms.date: 04/21/2021
---

> # Tutorial: Make HTTP requests in a .NET console application using C\#

Reviewer: Fine with me but I thought the guidelines recommended app for application?

Author: The VS UI has "Console Application", and [dotnet new](https://docs.microsoft.com/en-us/dotnet/core/tools/dotnet-new#arguments) has the same. The style guide does say ["most of the time use app"](https://styleguides.azurewebsites.net/Styleguide/Read?id=2700&topicid=32507) but also says "It's OK to use application in content for a technical audience".  I'll change to app.

   * Replaces the `Console.WriteLine` statement with a call to `ProcessRepositories` that uses the `await` keyword.

1. In the `Program` class, create a static instance of <xref:System.Net.Http.HttpClient> to handle requests and responses.

   ```csharp
   namespace WebAPIClient
   {
       class Program
       {
           private static readonly HttpClient client = new HttpClient();

           static async Task Main(string[] args)
           {
               //...
           }
       }
   }
   ```

1. In the `ProcessRepositories` method, call the GitHub endpoint that returns a list of all repositories under the .NET foundation organization:

   ```csharp
   private static async Task ProcessRepositories()
   {
       client.DefaultRequestHeaders.Accept.Clear();
       client.DefaultRequestHeaders.Accept.Add(
           new MediaTypeWithQualityHeaderValue("application/vnd.github.v3+json"));
       client.DefaultRequestHeaders.Add("User-Agent", ".NET Foundation Repository Reporter");
   
       var stringTask = client.GetStringAsync("https://api.github.com/orgs/dotnet/repos");
   
       var msg = await stringTask;
       Console.Write(msg);
   }
   ```

   This code:

>    * Sets up HTTP headers for all requests: an `Accept` header to accept JSON responses, and a `User-Agent` header. These headers are checked by the GitHub server code and are necessary to retrieve information from GitHub.

Reviewer: ```suggestion
   * Sets up HTTP headers for all requests: an [`Accept`](https://developer.mozilla.org/docs/Web/HTTP/Headers/Accept) header to accept JSON responses, and a `User-Agent` header. These headers are checked by the GitHub server code and are necessary to retrieve information from GitHub.
```

Reviewer: https://developer.mozilla.org/docs/Web/HTTP/Headers/User-Agent

Reviewer: I'd be tempted to split up the 2 header rather than and them. It's a long sentence.

   * Calls <xref:System.Net.Http.HttpClient.GetStringAsync(System.String)?displayProperty=nameWithType> to make a web request and retrieve the response. This method starts a task that makes the web request. When the request returns, the task reads the response stream and extracts the content from the stream. The body of the response is returned as a <xref:System.String>, which is available when the task completes.
   * Awaits the task for the response string and prints the response to the console.

1. Add two `using` directives at the top of the file:

   ```csharp
   using System.Net.Http;
   using System.Net.Http.Headers;
   ```

1. Build the application and run it.

   ```dotnetcli
   dotnet run
   ```

>    There is no build warning, because the `ProcessRepositories` now does contain an `await` operator.

Reviewer: ```suggestion
   There is no build warning  because the `ProcessRepositories` ~now~ contains an `await` operator.
```

   The output is a long display of JSON text.

## Deserialize the JSON Result

The following steps convert the JSON response into C# objects. You use the <xref:System.Text.Json.JsonSerializer?displayProperty=nameWithType> class to deserialize JSON into objects.

> 1. Create a file named *repo.cs* and put the following code in it. This code defines a class to represent the JSON object returned from the GitHub API:

Reviewer: Your version is fine. I write it like the following:
```suggestion
1. Create a file named *repo.cs* and add the following code:
```

Move this after
The preceding code defines a class to represent the JSON object returned from the GitHub API:

   ```csharp
   using System;
   
   namespace WebAPIClient
   {
       public class Repository
       {
           public string name { get; set; }
       }
   }
   ```

   You'll use this class to display a list of repository names.

   The JSON for a repository object contains dozens of properties,but only the `name` property will be deserialized. The serializer automatically ignores JSON properties for which there is no match in the target class. This feature makes it easier to create types that work with only a subset of fields in a large JSON packet.

>    The C# convention is to capitalize the first letter of property names, but the `name` property here starts with a lowercase letter because that matches exactly what's in the JSON. Later you'll see how to use C# property names that don't match the JSON property names.

Reviewer: https://docs.microsoft.com/en-us/dotnet/standard/design-guidelines/capitalization-conventions

1. Use the serializer to convert JSON into C# objects. Replace the call to
<xref:System.Net.Http.HttpClient.GetStringAsync(System.String)> in the `ProcessRepositories` method with the following lines:

   ```csharp
   var streamTask = client.GetStreamAsync("https://api.github.com/orgs/dotnet/repos");
   var repositories = await JsonSerializer.DeserializeAsync<List<Repository>>(await streamTask);
   ```

>    The new code calls <xref:System.Net.Http.HttpClient.GetStreamAsync(System.String)> instead of <xref:System.Net.Http.HttpClient.GetStringAsync(System.String)>. This serializer method uses a stream instead of a string as its source.

Reviewer: Perfectly written. I'd write it as:

```suggestion
   The updated code replaces <xref:System.Net.Http.HttpClient.GetStringAsync(System.String)> with <xref:System.Net.Http.HttpClient.GetStreamAsync(System.String)>. This serializer method uses a stream instead of a string as its source.
```

>    The first argument to <xref:System.Text.Json.JsonSerializer.DeserializeAsync%60%601(System.IO.Stream,System.Text.Json.JsonSerializerOptions,System.Threading.CancellationToken)?displayProperty=nameWithType> is an `await` expression. (The other two parameters are optional and are omitted in the code snippet.) `await` expressions can appear almost anywhere in your code, even though up to now, you've only seen them as part of an assignment statement.

Reviewer: ```suggestion
   The first argument to <xref:System.Text.Json.JsonSerializer.DeserializeAsync%60%601(System.IO.Stream,System.Text.Json.JsonSerializerOptions,System.Threading.CancellationToken)?displayProperty=nameWithType> is an `await` expression, the other two parameters are optional and are omitted in the code snippet. `await` expressions can appear almost anywhere in your code, even though up to now, you've only seen them as part of an assignment statement.
```

>    The `DeserializeAsync` method is *generic*, which means you supply type arguments for what kind of objects should be created from the JSON text. In this example, you're deserializing to a `List<Repository>`, which is another generic object, a <xref:System.Collections.Generic.List%601?displayProperty=nameWithType>. The `List<T>` class stores a collection of objects. The type argument declares the type of objects stored in the `List<T>`. The type argument is your `Repository` class, because the JSON text represents a collection of repository objects.

Reviewer: link to https://docs.microsoft.com/en-us/dotnet/csharp/programming-guide/generics/generic-methods or better https://docs.microsoft.com/en-us/dotnet/csharp/programming-guide/generics/

That's not what generic means, but generic allows - Right?

Author: >That's not what generic means, but generic allows - Right?

In this case it's more than just allows, as the type parameter is required.

   ```csharp
   var msg = await stringTask;
   Console.Write(msg);
   ```

   with the following code:

   ```csharp
   foreach (var repo in repositories)
       Console.WriteLine(repo.name);
   ```

1. Add the following `using` directives at the top of the file:

   ```csharp
   using System.Collections.Generic;
   using System.Text.Json;
   ```

1. Run the application.

   ```dotnetcli
   dotnet run
   ```

   The output is a list of the names of the repositories that are part of the .NET Foundation.

## Configure deserialization

1. In *repo.cs*, change the `name` property to `Name` and add a `[JsonPropertyName]` attribute to specify how this property appears in the JSON.

   ```csharp
   [JsonPropertyName("name")]
   public string Name { get; set; }
   ```

1. Add the <xref:System.Text.Json.Serialization> namespace to the `using` directives:

   ```csharp
   using System.Text.Json.Serialization;
   ```

1. In *Program.cs*, update the code to use the new capitalization of the `Name` property:

   ```csharp
   Console.WriteLine(repo.Name);
   ```

1. Run the application.

   The output is the same.

## Refactor the code

> The `ProcessRepositories` method can do the async work and return a collection of the repositories. Change that method to return `List<Repository>`, and move the code that writes the information into the `Main` method.

Reviewer: ```suggestion
The `ProcessRepositories` method does the async work and returns a collection of the repositories. Change that method to return `List<Repository>`, and move the code that writes the information into the `Main` method.
```

Author: Here present tense doesn't work because it's describing the rationale for the refactoring, it's not saying what the code already does.

Reviewer 2: I agree with Tom here.

1. Change the signature of `ProcessRepositories` to return a task whose result is a list of `Repository` objects:

   ```csharp
   private static async Task<List<Repository>> ProcessRepositories()
   ```

1. Return the repositories after processing the JSON response:

   ```csharp
   var streamTask = client.GetStreamAsync("https://api.github.com/orgs/dotnet/repos");
   var repositories = await JsonSerializer.DeserializeAsync<List<Repository>>(await streamTask);
   return repositories;
   ```

   The compiler generates the `Task<T>` object for the return value because you've marked this method as `async`.

1. Modify the `Main` method to capture the results and write each repository name to the console. The `Main` method now looks like this:

   ```csharp
   public static async Task Main(string[] args)
   {
       var repositories = await ProcessRepositories();
   
       foreach (var repo in repositories)
           Console.WriteLine(repo.Name);
   }
   ```

1. Run the application.

   The output is the same.

## Deserialize more properties

> The following steps add code to process a few more of the properties in the JSON packet that is received. You won't want to grab everything, but adding a few properties demonstrates more features of the C# language.

Reviewer: ```suggestion
The following steps add code to process more of the properties in the  received JSON packet. You probably won't want to process every property, but adding a few more demonstrates other features of C#.
```

1. Add the following properties to the `Repository` class definition:

   ```csharp
   [JsonPropertyName("description")]
   public string Description { get; set; }
   
   [JsonPropertyName("html_url")]
   public Uri GitHubHomeUrl { get; set; }
   
   [JsonPropertyName("homepage")]
   public Uri Homepage { get; set; }
   
   [JsonPropertyName("watchers")]
   public int Watchers { get; set; }
   ```

>    The <xref:System.Uri> and `int` types have built-in functionality to convert to and from string representation. So no extra code is needed to deserialize from JSON string format to those target types. If the JSON packet contains data that doesn't convert to a target type, the serialization action throws an exception.

Reviewer: ```suggestion
   The <xref:System.Uri> and `int` types have built-in functionality to convert to and from string representations. No extra code is needed to deserialize from JSON string format to those target types. If the JSON packet contains data that doesn't convert to a target type, the serialization action throws an exception.
```

To get a date and time represented in your time zone, you have to write a custom conversion method.

1. In *repo.cs*, add a `public` property for the UTC representation of the date and time and a `LastPush` `readonly` property that returns the date converted to local time:

   ```csharp
   [JsonPropertyName("pushed_at")]
   public DateTime LastPushUtc { get; set; }

   public DateTime LastPush => LastPushUtc.ToLocalTime();
   ```

   The `LastPush` property is defined using an *expression-bodied member* for the `get` accessor. There's no `set` accessor. Omitting the `set`  accessor is one way to define a *read-only* property in C#. (Yes, you can create *write-only* properties in C#, but their value is limited.)

1. Add another output statement in *Program.cs*:
again:

   ```csharp
   Console.WriteLine(repo.LastPush);
   ```

1. Run the application.

   The output includes the date and time of the last push to each repository.

## Next steps

Your version of the application should now match the [finished sample](https://github.com/dotnet/samples/tree/main/csharp/getting-started/console-webapiclient).

> This tutorial showed you how to make web requests, parse the results, and display properties of those results. You've seen some of the features of the C# language that support object-oriented techniques.

Reviewer: Not a fan of what you learned :(

---
title: "Tutorial: Make HTTP requests in a .NET console app using C#"
description: Learn how to make HTTP requests to a REST web service and deserialize JSON responses. This tutorial creates a .NET console and uses C#.
ms.date: 04/21/2021
---

# Tutorial: Make HTTP requests in a .NET console app using C\#

This tutorial builds an app that issues HTTP requests to a REST service on GitHub. The app reads information in JSON format and converts the JSON into C# objects. Converting from JSON to C# objects is known as *deserialization*.

The tutorial shows how to:

> [!div class="checklist"]
>
> * Send HTTP requests.
> * Deserialize JSON responses.
> * Configure deserialization with attributes.

> If you prefer to follow along with the [final sample](https://github.com/dotnet/samples/tree/main/csharp/getting-started/console-webapiclient) for this tutorial, you can download it. For download instructions, see [Samples and Tutorials](../../samples-and-tutorials/index.md#view-and-download-samples).

Reviewer 3: nit: Samples browser is more suitable when it comes to downloading single samples.

Reviewer 2: I'd agree. We could address this in a separate PR, as that would update the samples repo.
